text/3064-bitfield.md

```diff
@@ -0,0 +1,384 @@
+containing bit-fields.
+
+The intended behavior is that the layout is the same layout as produced by Clang
+when compiling the corresponding C struct for the same target. The corresponding
+C struct is constructed as follows:
+
+- Translate the struct to the corresponding C struct as if the `bitfield`
+  annotations were not present and as if `_` were a valid identifier for regular
+  fields.
+- For each field that has a `bitfield(N)` annotation in the Rust struct, append
+  `: N` to the declaration in the C struct.
+- For each field in the C struct whose name is `_`, delete the field name.
+
+The `&` and `&mut` operators cannot be applied to bit-fields.
+
+This implies that, in order to (mutably) access a bit-field, one must have a
+(mutable) reference to the struct. Therefore, no concurrent accesses of the
+struct are possible unless all accesses are immutable. Therefore, unlike in C,
+the compiler can implement access to bit-fields with any kind of load/write
+instruction, even if such an instruction overlaps the memory locations of other
+fields in the struct.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Bit-fields cannot be assigned without overflow checks.
+- The annotation feels somewhat far away from the type itself.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- Alternative: Using `u32: N` to specify the width of a bit-field. I've chosen
+  to use an annotation instead because it seems more likely to be accepted.
+
+- Alternative: Using a dedicated type instead of an annotation:
+
+    ```rust
+    #[repr(C)]
+    struct X {
+        a: BitField<u32, 6>,
+    }
+    ```
+
+    where `BitField` would be some kind of lang item that ensures that
+
+    - the address of a field with such a type cannot be taken
+    - the type and width parameters are correct.
+
+    This type would then have the following API:
+
+    ```rust
+    impl<T, const N: usize> BitField<T, N> {
+        fn new(t: T) -> Self;
+        fn wrapping_new(t: T) -> Self;
+        fn get(self) -> T;
+        fn set(&mut self, t: T);
+        fn wrapping_add(self, t: T) -> Self;
+        ...
+    }
+
+    impl<T, const N: usize> AddAssign<T> for BitField<T, N> {
+    }
+    ```
+
+    The layout of `BitField<T, N>` would be the layout of `T` when used outside
+    a struct.
+
+    This could then be used like so:
+
+    ```rust
+    fn f(x: X) {
+        x.a.set(1);
+        x.a += 1;
+        x.a = x.a.wrapping_add(12);
+        println!("{}", x.a.get());
+    }
+    ```
+
+    There is some magic happening here:
+
+    - `x.a += 1` cannot be written as `AddAssign::add(&mut x.a, 1)` because the
+      address of `x.a` cannot be taken. The compiler would figure it out when
+      the lhs is a struct field.
+    - When the assignment happens, the compiler writes the bits on the rhs to
+      the correct positions in the struct.
+
+    This is a more structured solution but would need to be fleshed out.
```

Comment:
I agree both that this might be a good long-term solution (and bit-sized types would be useful for other purposes), and that this should not block the implementation of bitfields.

If, in the future, we add bit-sized types, we can specify how to put them in a `repr(C)` struct and have them behave like a bitfield. That should not stop us from providing the most straightforward implementation of bitfields without introducing bit-sized types.